crates/runtime/signalhandlers/Trampolines.cpp

```diff
@@ -1,12 +1,69 @@
+#include <unistd.h>
 #include <setjmp.h>
+#include <stdlib.h>
+#include <sys/mman.h>
 
 #include "SignalHandlers.hpp"
 
+// The size of the sigaltstack (not including the guard, which will be added).
+// Make this large enough to run our signal handlers.
+static const size_t sigaltstack_size = 4 * SIGSTKSZ;
+
+// A utility to register a new sigaltstack.
+namespace {
+  static thread_local class SigAltStack {
+    size_t guard_size;
+    size_t sigaltstack_alloc_size;
+    stack_t new_stack;
+
+  public:
+    SigAltStack();
+    ~SigAltStack();
+  } thread_sigaltstack;
+}
+
+SigAltStack::SigAltStack()
+  : guard_size(sysconf(_SC_PAGESIZE))
+  , sigaltstack_alloc_size(guard_size + sigaltstack_size)
+{
+  // Allocate memory.
+  void *ptr = mmap(NULL, sigaltstack_alloc_size, PROT_NONE,
+                   MAP_PRIVATE | MAP_ANON, -1, 0);
+  if (ptr == MAP_FAILED)
+    RaiseOOMTrap();
+
+  // Prepare the stack, register it, and sanity check the old stack.
+  void *stack_ptr = (void *)((uintptr_t)ptr + guard_size);
+  new_stack = (stack_t) { stack_ptr, 0, sigaltstack_size };
+  stack_t old_stack;
+  if (mprotect(stack_ptr, sigaltstack_size, PROT_READ | PROT_WRITE) != 0 ||
+      sigaltstack(&new_stack, &old_stack) != 0 ||
+      old_stack.ss_flags != 0 ||
+      old_stack.ss_size > sigaltstack_size)
```

Comment:
Could this test be done before we start mmap'ing? We could check to see if the sigaltstack size is already big enough or if it's disabled then we we can act before we mmap our own and set it.